File: src/lib/whiteboardGeometry.test.ts
import { strict as assert } from 'node:assert'
import test from 'node:test'
import {
  BOARD, NOTE_HEIGHT, NOTE_WIDTH, anchorOf, clampToBoard, clampView, defaultControls,
  linkPath, matchNotes, minimapViewport, noteAt, panViewByBoardDelta, sidesBetween, toBoard, viewCentredOn, viewFromMinimapPoint,
} from './whiteboardGeometry.ts'

const note = (id: string, x: number, y: number, text = '') => ({ id, x, y, text })

test('a point past an edge is brought back onto the board', () => {
  assert.deepEqual(clampToBoard({ x: -50, y: -50 }), { x: 0, y: 0 })
  assert.deepEqual(clampToBoard({ x: 99999, y: 99999 }), { x: BOARD.width, y: BOARD.height })
})

test('a note is clamped by its far edge, not its corner', () => {
  const placed = clampToBoard({ x: BOARD.width - 10, y: 0 }, { width: NOTE_WIDTH, height: NOTE_HEIGHT })
  assert.equal(placed.x, BOARD.width - NOTE_WIDTH)
})

test('panning stops at the board edge rather than running on into nothing', () => {
  const viewport = { width: 1000, height: 700 }
  assert.equal(clampView({ x: 500, y: 300, scale: 1 }, viewport).x, 0)
  const far = clampView({ x: -99999, y: -99999, scale: 1 }, viewport)
  assert.equal(far.x, viewport.width - BOARD.width)
  assert.equal(far.y, viewport.height - BOARD.height)
})

test('a board smaller than its window is centred, not cornered', () => {
  const viewport = { width: 1000, height: 700 }
  const view = clampView({ x: -400, y: -400, scale: 0.05 }, viewport)
  assert.equal(view.x, (1000 - BOARD.width * 0.05) / 2)
  assert.equal(view.y, (700 - BOARD.height * 0.05) / 2)
})

test('a view already inside the board is left alone', () => {
  const view = { x: -1200, y: -800, scale: 1 }
  assert.deepEqual(clampView(view, { width: 1000, height: 700 }), view)
})

test('screen pixels map back to board coordinates through the view', () => {
  const view = { x: -200, y: -100, scale: 2 }
  assert.deepEqual(toBoard({ x: 400, y: 300 }, view), { x: 300, y: 200 })
})

test('connectors leave the middle of an edge', () => {
  const subject = note('a', 100, 200)
  assert.deepEqual(anchorOf(subject, 'start'), { x: 100, y: 200 + NOTE_HEIGHT / 2 })
  assert.deepEqual(anchorOf(subject, 'end'), { x: 100 + NOTE_WIDTH, y: 200 + NOTE_HEIGHT / 2 })
})

test('a note to the right is joined end-to-start, and one to the left the other way', () => {
  const left = note('a', 0, 0)
  const right = note('b', 600, 0)
  assert.deepEqual(sidesBetween(left, right), { from: 'end', to: 'start' })
  assert.deepEqual(sidesBetween(right, left), { from: 'start', to: 'end' })
  // Stacked notes read left-to-right, so the tie goes to the forward direction.
  assert.deepEqual(sidesBetween(left, note('c', 0, 400)), { from: 'end', to: 'start' })
})

test('the default bend eases horizontally out of both ends', () => {
  const [c1, c2] = defaultControls({ x: 0, y: 50 }, { x: 400, y: 250 })
  assert.equal(c1.y, 50)
  assert.equal(c2.y, 250)
  assert.equal(c1.x, 200)
  assert.equal(c2.x, 200)
})

test('a very short connector still bends by a visible minimum', () => {
  const [c1] = defaultControls({ x: 0, y: 0 }, { x: 10, y: 0 })
  assert.equal(c1.x, 40)
})

test('a bent connector keeps the control points it was given', () => {
  const bent = linkPath({ x: 0, y: 0 }, { x: 100, y: 0 }, [{ x: 10, y: -80 }, { x: 90, y: 80 }])
  assert.equal(bent.d, 'M 0 0 C 10 -80, 90 80, 100 0')
  assert.deepEqual(bent.controls, [{ x: 10, y: -80 }, { x: 90, y: 80 }])
})

test('the topmost note wins when two overlap', () => {
  const under = note('under', 0, 0)
  const over = note('over', 20, 20)
  assert.equal(noteAt([under, over], { x: 40, y: 40 })?.id, 'over')
})

test('a point outside every note hits nothing', () => {
  assert.equal(noteAt([note('a', 0, 0)], { x: 900, y: 900 }), null)
})

test('search matches text, case-insensitively, and skips notes without it', () => {
  const notes = [note('a', 0, 0, 'Preload and afterload'), note('b', 0, 0, 'Murmurs'), note('c', 0, 0, '')]
  assert.deepEqual(matchNotes(notes, 'PRELOAD').map((n) => n.id), ['a'])
  assert.deepEqual(matchNotes(notes, ''), [])
})

test('hits are stepped through in reading order, not the order they were made', () => {
  const notes = [
    note('bottom', 10, 800, 'aortic'),
    note('top-right', 900, 10, 'aortic'),
    note('top-left', 10, 10, 'aortic'),
  ]
  assert.deepEqual(matchNotes(notes, 'aortic').map((n) => n.id), ['top-left', 'top-right', 'bottom'])
})

test('centring on a point puts it mid-viewport, and stays on the board', () => {
  const viewport = { width: 1000, height: 700 }
  const view = viewCentredOn({ x: 2000, y: 1500 }, viewport, 1)
  assert.equal(view.x, 500 - 2000)
  assert.equal(view.y, 350 - 1500)
  // A point near the corner cannot be centred without leaving the board.
  assert.equal(viewCentredOn({ x: 0, y: 0 }, viewport, 1).x, 0)
})

test('minimap viewport is computed in board coordinates without rounding', () => {
  const view = { x: -123.456, y: -78.9, scale: 1.25 }
  const viewport = minimapViewport(view, { width: 1440.5, height: 812.25 }, { width: 190, height: 112 })
  assert.equal(viewport.x, (-view.x / view.scale) * (112 / BOARD.height))
  assert.equal(viewport.y, (-view.y / view.scale) * (112 / BOARD.height))
  assert.equal(viewport.width, (1440.5 / view.scale) * (112 / BOARD.height))
  assert.equal(viewport.height, (812.25 / view.scale) * (112 / BOARD.height))
})

test('clicking the minimap centres that board point in the main viewport', () => {
  const viewport = { width: 1000, height: 700 }
  const view = viewFromMinimapPoint({ x: 95, y: 56 }, viewport, { width: 190, height: 112 }, { x: 0, y: 0, scale: 1 })
  assert.equal(view.x, 500 - (95 / (112 / BOARD.height)))
  assert.equal(view.y, 350 - (56 / (112 / BOARD.height)))
})

test('keyboard minimap panning moves in board units and respects clamps', () => {
  const viewport = { width: 1000, height: 700 }
  const moved = panViewByBoardDelta({ x: -500, y: -500, scale: 2 }, { x: 12.5, y: -20.25 }, viewport)
  assert.equal(moved.x, -525)
  assert.equal(moved.y, -459.5)
  assert.equal(panViewByBoardDelta({ x: 0, y: 0, scale: 1 }, { x: -100, y: 0 }, viewport).x, 0)
})

File: android/app/src/test/java/com/synapse/app/core/whiteboard/WhiteboardGeometryTest.kt
package com.synapse.app.core.whiteboard

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

/**
 * A verbatim port of every vector in `src/lib/whiteboardGeometry.test.ts` — the
 * geometry must stay numerically identical across web, iOS and Android, and this
 * is the test that proves it for Android. Test names mirror the TS ones.
 */
class WhiteboardGeometryTest {

    private fun note(id: String, x: Double, y: Double, text: String = ""): NoteBox = NoteBox(id, x, y, text)

    @Test
    fun `a point past an edge is brought back onto the board`() {
        assertEquals(Point(0.0, 0.0), clampToBoard(Point(-50.0, -50.0)))
        assertEquals(Point(Board.WIDTH, Board.HEIGHT), clampToBoard(Point(99999.0, 99999.0)))
    }

    @Test
    fun `a note is clamped by its far edge, not its corner`() {
        val placed = clampToBoard(Point(Board.WIDTH - 10, 0.0), Size(NOTE_WIDTH, NOTE_HEIGHT))
        assertEquals(Board.WIDTH - NOTE_WIDTH, placed.x, 0.0)
    }

    @Test
    fun `panning stops at the board edge rather than running on into nothing`() {
        val viewport = Size(1000.0, 700.0)
        assertEquals(0.0, clampView(BoardView(500.0, 300.0, 1.0), viewport).x, 0.0)
        val far = clampView(BoardView(-99999.0, -99999.0, 1.0), viewport)
        assertEquals(viewport.width - Board.WIDTH, far.x, 0.0)
        assertEquals(viewport.height - Board.HEIGHT, far.y, 0.0)
    }

    @Test
    fun `a board smaller than its window is centred, not cornered`() {
        val viewport = Size(1000.0, 700.0)
        val view = clampView(BoardView(-400.0, -400.0, 0.05), viewport)
        assertEquals((1000 - Board.WIDTH * 0.05) / 2, view.x, 1e-9)
        assertEquals((700 - Board.HEIGHT * 0.05) / 2, view.y, 1e-9)
    }

    @Test
    fun `a view already inside the board is left alone`() {
        val view = BoardView(-1200.0, -800.0, 1.0)
        assertEquals(view, clampView(view, Size(1000.0, 700.0)))
    }

    @Test
    fun `screen pixels map back to board coordinates through the view`() {
        val view = BoardView(-200.0, -100.0, 2.0)
        assertEquals(Point(300.0, 200.0), toBoard(Point(400.0, 300.0), view))
    }

    @Test
    fun `connectors leave the middle of an edge`() {
        val subject = note("a", 100.0, 200.0)
        assertEquals(Point(100.0, 200.0 + NOTE_HEIGHT / 2), anchorOf(subject, Side.Start))
        assertEquals(Point(100.0 + NOTE_WIDTH, 200.0 + NOTE_HEIGHT / 2), anchorOf(subject, Side.End))
    }

    @Test
    fun `a note to the right is joined end-to-start, and one to the left the other way`() {
        val left = note("a", 0.0, 0.0)
        val right = note("b", 600.0, 0.0)
        assertEquals(SidePair(Side.End, Side.Start), sidesBetween(left, right))
        assertEquals(SidePair(Side.Start, Side.End), sidesBetween(right, left))
        // Stacked notes read left-to-right, so the tie goes to the forward direction.
        assertEquals(SidePair(Side.End, Side.Start), sidesBetween(left, note("c", 0.0, 400.0)))
    }

    @Test
    fun `the default bend eases horizontally out of both ends`() {
        val (c1, c2) = defaultControls(Point(0.0, 50.0), Point(400.0, 250.0))
        assertEquals(50.0, c1.y, 0.0)
        assertEquals(250.0, c2.y, 0.0)
        assertEquals(200.0, c1.x, 0.0)
        assertEquals(200.0, c2.x, 0.0)
    }

    @Test
    fun `a very short connector still bends by a visible minimum`() {
        val (c1, _) = defaultControls(Point(0.0, 0.0), Point(10.0, 0.0))
        assertEquals(40.0, c1.x, 0.0)
    }

    @Test
    fun `a bent connector keeps the control points it was given`() {
        val bent = linkPath(Point(0.0, 0.0), Point(100.0, 0.0), Point(10.0, -80.0) to Point(90.0, 80.0))
        assertEquals("M 0 0 C 10 -80, 90 80, 100 0", bent.d)
        assertEquals(Point(10.0, -80.0) to Point(90.0, 80.0), bent.controls)
    }

    @Test
    fun `the topmost note wins when two overlap`() {
        val under = note("under", 0.0, 0.0)
        val over = note("over", 20.0, 20.0)
        assertEquals("over", noteAt(listOf(under, over), Point(40.0, 40.0))?.id)
    }

    @Test
    fun `a point outside every note hits nothing`() {
        assertNull(noteAt(listOf(note("a", 0.0, 0.0)), Point(900.0, 900.0)))
    }

    @Test
    fun `search matches text, case-insensitively, and skips notes without it`() {
        val notes = listOf(
            note("a", 0.0, 0.0, "Preload and afterload"),
            note("b", 0.0, 0.0, "Murmurs"),
            note("c", 0.0, 0.0, ""),
        )
        assertEquals(listOf("a"), matchNotes(notes, "PRELOAD").map { it.id })
        assertEquals(emptyList<String>(), matchNotes(notes, "").map { it.id })
    }

    @Test
    fun `hits are stepped through in reading order, not the order they were made`() {
        val notes = listOf(
            note("bottom", 10.0, 800.0, "aortic"),
            note("top-right", 900.0, 10.0, "aortic"),
            note("top-left", 10.0, 10.0, "aortic"),
        )
        assertEquals(listOf("top-left", "top-right", "bottom"), matchNotes(notes, "aortic").map { it.id })
    }

    @Test
    fun `centring on a point puts it mid-viewport, and stays on the board`() {
        val viewport = Size(1000.0, 700.0)
        val view = viewCentredOn(Point(2000.0, 1500.0), viewport, 1.0)
        assertEquals(500.0 - 2000.0, view.x, 0.0)
        assertEquals(350.0 - 1500.0, view.y, 0.0)
        // A point near the corner cannot be centred without leaving the board.
        assertEquals(0.0, viewCentredOn(Point(0.0, 0.0), viewport, 1.0).x, 0.0)
    }

    @Test
    fun `minimap viewport is computed in board coordinates without rounding`() {
        val view = BoardView(-123.456, -78.9, 1.25)
        val viewport = minimapViewport(view, Size(1440.5, 812.25), Size(190.0, 112.0))
        assertEquals((-view.x / view.scale) * (112.0 / Board.HEIGHT), viewport.x, 1e-9)
        assertEquals((-view.y / view.scale) * (112.0 / Board.HEIGHT), viewport.y, 1e-9)
        assertEquals((1440.5 / view.scale) * (112.0 / Board.HEIGHT), viewport.width, 1e-9)
        assertEquals((812.25 / view.scale) * (112.0 / Board.HEIGHT), viewport.height, 1e-9)
    }

    @Test
    fun `clicking the minimap centres that board point in the main viewport`() {
        val viewport = Size(1000.0, 700.0)
        val view = viewFromMinimapPoint(Point(95.0, 56.0), viewport, Size(190.0, 112.0), BoardView(0.0, 0.0, 1.0))
        assertEquals(500.0 - (95.0 / (112.0 / Board.HEIGHT)), view.x, 1e-9)
        assertEquals(350.0 - (56.0 / (112.0 / Board.HEIGHT)), view.y, 1e-9)
    }

    @Test
    fun `keyboard minimap panning moves in board units and respects clamps`() {
        val viewport = Size(1000.0, 700.0)
        val moved = panViewByBoardDelta(BoardView(-500.0, -500.0, 2.0), Point(12.5, -20.25), viewport)
        assertEquals(-525.0, moved.x, 0.0)
        assertEquals(-459.5, moved.y, 0.0)
        assertEquals(0.0, panViewByBoardDelta(BoardView(0.0, 0.0, 1.0), Point(-100.0, 0.0), viewport).x, 0.0)
    }
}
